### ISSUE TITLE
[firebase_messaging] FirebaseMessaging.onBackgroundMessage not called in iOS after restarting the device or a few hours of app being closed

### ISSUE THREAD

**Issue by yoniker** [firebase/flutterfire#7785]
## Bug report

FCM is working on my physical iPhone 13 **as long as there the phone wasn't restarted,** (or otherwise if I used the app in the last couple of hours).
After restarting the phone, the top level method I registered with onBackgroundMessage isn't being called. 
I wasn't able to debug it through log messages (see https://github.com/flutter/flutter/issues/96038#issuecomment-1003972152 which I'm not sure why it was closed, as Flutter logs should be Flutter's responsibility and not a plugin's responsibility)
so in the first line of that registered callback method I'm writing to a file, and when there's a notification that file is being changed and when there's no notification that file is being unchanged, meaning that the top level method wasn't even called (that explains why there's no notification produced).

1. Follow the instructions here to setup Firebase on flutter https://firebase.flutter.dev/docs/overview
2. and the instructions here to setup FCM in general https://firebase.flutter.dev/docs/messaging/overview/
3. and Finally the instructions here here  https://firebase.flutter.dev/docs/messaging/apple-integration (with your apple developer account) to setup FCM via the APN service
4. Install the app and run the app at least once (send FCM token to your server or otherwise write it down so you can manually try to send a message via the firebase console). 
5. Restart the physical iPhone (power off and then power back on) and try to send a FCM message via your server or through firebase console to the phone

### Expected behavior

the top level method I registered with onBackgroundMessage is being called 

## Additional context

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>

</summary>

```
Yonis-MacBook-Pro:Swiper yoni$ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.5.3, on macOS 12.0.1 21A559 darwin-x64, locale en-IL)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 2020.3)
[✓] Connected device (1 available)

• No issues found!
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
Yonis-MacBook-Pro:Swiper yoni$ flutter pub deps -- --style=compact
Dart SDK 2.14.4
Flutter SDK 2.5.3
betabeta 0.4.1+0.4.1

dependencies:
- auth_buttons 2.0.3 [flutter flutter_svg google_fonts]
- cached_network_image 3.2.0 [flutter flutter_cache_manager octo_image cached_network_image_platform_interface cached_network_image_web]
- carousel_slider 4.0.0 [flutter]
- connectivity_plus 2.1.0 [flutter connectivity_plus_platform_interface connectivity_plus_linux connectivity_plus_macos connectivity_plus_web connectivity_plus_windows]
- cupertino_icons 1.0.4
- firebase_core 1.10.6 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_messaging 11.2.4 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- flutter_cache_manager 3.3.0 [clock collection file flutter http path path_provider pedantic rxdart sqflite uuid]
- flutter_chat_ui 1.5.3 [flutter diffutil_dart equatable flutter_chat_types flutter_link_previewer intl meta photo_view]
- flutter_facebook_auth 3.5.7 [flutter flutter_facebook_auth_platform_interface flutter_facebook_auth_web]
- flutter_local_notifications 9.2.0 [clock flutter flutter_local_notifications_linux flutter_local_notifications_platform_interface timezone]
- flutter_spinkit 5.1.0 [flutter]
- flutter_svg 1.0.0 [flutter meta path_drawing vector_math xml]
- font_awesome_flutter 9.2.0 [flutter]
- fuzzy 0.4.0-nullsafety.0 [latinize]
- get 4.6.1 [flutter]
- hive 2.0.5 [meta crypto]
- hive_flutter 1.1.0 [flutter hive path_provider path]
- http 0.13.4 [async http_parser meta path]
- image 3.1.0 [archive meta xml]
- image_picker 0.8.4+4 [flutter flutter_plugin_android_lifecycle image_picker_for_web image_picker_platform_interface]
- page_indicator 0.4.1 [flutter]
- path_provider 2.0.8 [flutter path_provider_android path_provider_ios path_provider_linux path_provider_macos path_provider_platform_interface path_provider_windows]
- photo_view 0.13.0 [flutter]
- provider 6.0.2 [collection flutter nested]
- reorderables 0.4.2 [flutter]
- shared_preferences 2.0.11 [flutter meta shared_preferences_android shared_preferences_ios shared_preferences_linux shared_preferences_macos shared_preferences_platform_interface shared_preferences_web shared_preferences_windows]
- speech_bubble 0.0.9 [flutter]
- tcard 1.3.5 [flutter]
- trotter 2.0.0-dev.1
- tuple 2.0.0 [quiver]

dev dependencies:
- build_runner 2.1.7 [args async analyzer build build_config build_daemon build_resolvers build_runner_core code_builder collection crypto dart_style frontend_server_client glob graphs http_multi_server io js logging meta mime package_config path pool pub_semver pubspec_parse shelf shelf_web_socket stack_trace stream_transform timing watcher web_socket_channel yaml]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]
- hive_generator 1.1.1 [build source_gen hive analyzer source_helper]

transitive dependencies:
- _fe_analyzer_shared 31.0.0 [meta]
- analyzer 2.8.0 [_fe_analyzer_shared cli_util collection convert crypto glob meta package_config path pub_semver source_span watcher yaml]
- archive 3.1.6 [crypto path]
- args 2.3.0
- async 2.8.1 [collection meta]
- boolean_selector 2.1.0 [source_span string_scanner]
- build 2.2.1 [analyzer async convert crypto glob logging meta path]
- build_config 1.0.0 [checked_yaml json_annotation path pubspec_parse yaml]
- build_daemon 3.0.1 [built_collection built_value http_multi_server logging path pool shelf shelf_web_socket stream_transform watcher web_socket_channel]
- build_resolvers 2.0.6 [analyzer async build crypto graphs logging path package_config pool pub_semver stream_transform yaml]
- build_runner_core 7.2.3 [async build build_config build_resolvers collection convert crypto glob graphs json_annotation logging meta path package_config pool timing watcher yaml]
- built_collection 5.1.1
- built_value 8.1.3 [built_collection collection fixnum meta]
- cached_network_image_platform_interface 1.0.0 [flutter flutter_cache_manager]
- cached_network_image_web 1.0.1 [flutter flutter_cache_manager cached_network_image_platform_interface]
- characters 1.1.0
- charcode 1.3.1
- checked_yaml 2.0.1 [json_annotation source_span yaml]
- cli_util 0.3.5 [meta path]
- clock 1.1.0
- code_builder 4.1.0 [built_collection built_value collection matcher meta]
- collection 1.15.0
- connectivity_plus_linux 1.1.1 [flutter connectivity_plus_platform_interface meta nm]
- connectivity_plus_macos 1.2.1 [connectivity_plus_platform_interface flutter]
- connectivity_plus_platform_interface 1.1.1 [flutter meta plugin_platform_interface]
- connectivity_plus_web 1.1.0+1 [connectivity_plus_platform_interface flutter_web_plugins flutter]
- connectivity_plus_windows 1.2.0 [connectivity_plus_platform_interface flutter]
- convert 3.0.1 [typed_data]
- cross_file 0.3.2 [flutter js meta]
- crypto 3.0.1 [collection typed_data]
- csslib 0.17.1 [source_span]
- dart_style 2.2.1 [analyzer args path pub_semver source_span]
- dbus 0.6.6 [args ffi meta xml]
- diffutil_dart 3.0.0
- equatable 2.0.3 [collection meta]
- fake_async 1.2.0 [clock collection]
- ffi 1.1.2
- file 6.1.2 [meta path]
- firebase_core_platform_interface 4.2.3 [collection flutter meta plugin_platform_interface]
- firebase_core_web 1.5.3 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_messaging_platform_interface 3.1.4 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 2.2.5 [firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- fixnum 1.0.0
- flutter_blurhash 0.6.0 [flutter meta pedantic]
- flutter_chat_types 3.2.1 [equatable json_annotation meta]
- flutter_facebook_auth_platform_interface 2.7.1 [flutter plugin_platform_interface]
- flutter_facebook_auth_web 2.6.0+2 [flutter flutter_web_plugins js flutter_facebook_auth_platform_interface]
- flutter_link_previewer 2.6.1 [flutter flutter_chat_types flutter_linkify html http linkify meta url_launcher]
- flutter_linkify 5.0.2 [flutter linkify]
- flutter_local_notifications_linux 0.4.1+1 [flutter flutter_local_notifications_platform_interface dbus path xdg_directories]
- flutter_local_notifications_platform_interface 5.0.0 [flutter plugin_platform_interface]
- flutter_plugin_android_lifecycle 2.0.5 [flutter]
- flutter_web_plugins 0.0.0 [flutter js characters collection meta typed_data vector_math]
- frontend_server_client 2.1.2 [async path]
- glob 2.0.2 [async collection file path string_scanner]
- google_fonts 2.2.0 [flutter http path_provider crypto]
- graphs 2.1.0 [collection]
- html 0.15.0 [csslib source_span]
- http_multi_server 3.0.1 [async]
- http_parser 4.0.0 [charcode collection source_span string_scanner typed_data]
- image_picker_for_web 2.1.4 [flutter flutter_web_plugins image_picker_platform_interface meta]
- image_picker_platform_interface 2.4.1 [flutter http meta plugin_platform_interface cross_file]
- intl 0.17.0 [clock path]
- io 1.0.3 [meta path string_scanner]
- js 0.6.3
- json_annotation 4.4.0 [meta]
- latinize 0.1.0-nullsafety.0
- linkify 4.1.0
- logging 1.0.2
- matcher 0.12.10 [stack_trace]
- meta 1.7.0
- mime 1.0.1
- nested 1.0.0 [flutter]
- nm 0.4.2 [dbus]
- octo_image 1.0.1 [flutter flutter_blurhash]
- package_config 2.0.2 [path]
- path 1.8.0
- path_drawing 1.0.0 [vector_math meta path_parsing flutter]
- path_parsing 1.0.0 [vector_math meta]
- path_provider_android 2.0.11 [flutter path_provider_platform_interface]
- path_provider_ios 2.0.7 [flutter path_provider_platform_interface]
- path_provider_linux 2.1.4 [ffi flutter meta path path_provider_platform_interface xdg_directories]
- path_provider_macos 2.0.4 [flutter meta path_provider_platform_interface]
- path_provider_platform_interface 2.0.1 [flutter meta platform plugin_platform_interface]
- path_provider_windows 2.0.4 [ffi flutter meta path path_provider_platform_interface win32]
- pedantic 1.11.1
- petitparser 4.4.0 [meta]
- platform 3.1.0
- plugin_platform_interface 2.0.2 [meta]
- pool 1.5.0 [async stack_trace]
- process 4.2.4 [file path platform]
- pub_semver 2.1.0 [collection meta]
- pubspec_parse 1.2.0 [checked_yaml collection json_annotation pub_semver yaml]
- quiver 3.0.1+1 [matcher]
- rxdart 0.27.3
- shared_preferences_android 2.0.9 [flutter meta shared_preferences_platform_interface]
- shared_preferences_ios 2.0.8 [flutter meta shared_preferences_platform_interface]
- shared_preferences_linux 2.0.3 [file flutter meta path path_provider_linux shared_preferences_platform_interface]
- shared_preferences_macos 2.0.2 [flutter shared_preferences_platform_interface]
- shared_preferences_platform_interface 2.0.0 [flutter]
- shared_preferences_web 2.0.2 [flutter flutter_web_plugins meta shared_preferences_platform_interface]
- shared_preferences_windows 2.0.3 [file flutter meta path path_provider_platform_interface path_provider_windows shared_preferences_platform_interface]
- shelf 1.2.0 [async collection http_parser path stack_trace stream_channel]
- shelf_web_socket 1.0.1 [shelf stream_channel web_socket_channel]
- sky_engine 0.0.99
- source_gen 1.2.1 [analyzer async build dart_style glob meta path source_span yaml]
- source_helper 1.3.1 [analyzer collection source_gen]
- source_span 1.8.1 [collection path term_glyph]
- sqflite 2.0.1 [flutter sqflite_common path]
- sqflite_common 2.0.1+1 [synchronized path meta]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- stream_transform 2.0.0
- string_scanner 1.1.0 [charcode source_span]
- synchronized 3.0.0
- term_glyph 1.2.0
- test_api 0.4.2 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph matcher]
- timezone 0.8.0 [path]
- timing 1.0.0 [json_annotation]
- typed_data 1.3.0 [collection]
- url_launcher 6.0.17 [flutter meta url_launcher_android url_launcher_ios url_launcher_linux url_launcher_macos url_launcher_platform_interface url_launcher_web url_launcher_windows]
- url_launcher_android 6.0.13 [flutter meta url_launcher_platform_interface]
- url_launcher_ios 6.0.13 [flutter meta url_launcher_platform_interface]
- url_launcher_linux 2.0.2 [flutter]
- url_launcher_macos 2.0.2 [flutter]
- url_launcher_platform_interface 2.0.4 [flutter plugin_platform_interface]
- url_launcher_web 2.0.5 [flutter flutter_web_plugins meta url_launcher_platform_interface]
- url_launcher_windows 2.0.2 [flutter]
- uuid 3.0.5 [crypto]
- vector_math 2.1.0
- watcher 1.0.1 [async path]
- web_socket_channel 2.1.0 [async crypto stream_channel]
- win32 2.3.3 [ffi]
- xdg_directories 0.2.0 [meta path process]
- xml 5.3.1 [collection meta petitparser]
- yaml 3.1.0 [collection source_span string_scanner]
```

</details>

---

**Comment by darshankawar** [comment]
@yoniker 


> 5\. Restart the app on the physical iPhone and try to send a FCM message via your server or through firebase console to the phone

By restart, do you mean killing / terminating the app and then trying to open it again ? If so, are you terminating it with a swipe ?

For your reference, check below pre-requisite from the document and see if it helps:

```
The application must have opened at least once (to allow for registration with FCM).

On iOS, if the user swipes away the application from app Switcher, it must be manually reopened again for background messages to start working again.
```

**Comment by yoniker** [comment]
Oops, my mistake, I meant to write:
 restart the physical phone, and that's after using the app at least once.

So the server at that point has the FCM client token, but there's no notification.

Also, no notification if I use the app once, and wait "enough time" (I didn't test exactly how long, it's a few hours).

Finally (slightly related to the issue),regarding the documentation I'm curious:
On whatsapp, if the user swipes away the application away then there's still a notification.
How do I achieve the same behavior with FCM?

**Comment by darshankawar** [comment]
> How do I achieve the same behavior with FCM?

I am not completely sure how to achieve that after swiping away the app. On iOS atleast, we need to manually open the app.

Regarding your original case, curious to know the scenario wherein you need to restart a device and then send a notification ? Or that probably just occured to you while testing something ?

Also, other part of same query (ie, few hours of app being closed), there's a similar issue but for Android, https://github.com/FirebaseExtended/flutterfire/issues/6956, maybe you can take a look and see if that's same case as yours ?

**Comment by yoniker** [comment]
First of all, regarding #6956, sounds like a similar issue, I'll test it and let you know (I didn't notice any issues on Android until now by the way). I installed my app for this purpose on a physical Android device and will replace this text with the test results for your convenience. (by definition I have to check only after a few hours at least).

Are you curious about a scenario in which a device restarts and notification needs to be sent?! I probably misunderstand you, isn't this the most common use case for any app which includes notifications?  (Whatsapp,Tinder,Bumble,Facebook messenger etc). 
Most phones are restarted or powered off and back on when the battery ran low or ran out...
And all of the apps mentioned above want to deliver a message (after the app was used before restarting) even if the user didn't open the app after restarting (or power-off power-on) the phone. This is the expected behavior.
I found the issue by testing it.

Are you suggesting that a professional app with notifications (examples mentioned above) cannot be implemented using FCM?! An expectation that the user will open the app again after swiping away shouldn't exist - think about it from a user's point of view...

**Comment by darshankawar** [comment]
> Are you suggesting that a professional app with notifications (examples mentioned above) cannot be implemented using FCM?!

I am not suggesting that. I said `I am not completely sure how to achieve that after swiping away the app. On iOS atleast, we need to manually open the app.`

I am keeping this issue open for further input from the team, as I am not sure how to debug more on this scenario.

/cc @russellwheatley

**Comment by dimitristoz** [comment]
Hello @darshankawar , I am facing a similar issue only on IOS. I am sending a silent background notifications (with only data included), I use contentAvailable as true, and priority to high. I am facing an issue when the app is killed. It looks like randomly the token get unregistered. I am not able to identify when this happens because sometimes works properly after installing the app but sometimes after the kill I am facing this issue. I am using firebase_messaging 11.2.5,  and firebase_core: 1.11.0.

**Comment by russellwheatley** [comment]
Hello everyone, we’ve reviewed the issues opened for iOS background messages and we want to make our position clear.

Android and iOS handle background messaging differently. The decision made by the iOS operating system whether messages reaches the relevant iOS event handler (and subsequently received by the Dart background messaging handler) is based on a number of criteria, such as: CPU usage, priority level (data-only messages are considered “low priority” by iOS, android does not), battery level, amount of messages being received by the app, background/terminated application state, etc. This is a fundamental difference between the platforms, and you need to be aware of them when designing your application.

One solution suggested for iOS data-only messages unreliability has been to add a NotificationExtension. But this will still not create parity with Android as you will still have to include a notification with your message (therefore not data-only). It will still render the original notification if you do not update/mutate the notification that comes through the system in a timely manner.  We are not currently considering this as a solution due to these limitations.

We also will not support non-FCM messages from third party packages. We specifically only support messages received from the Firebase APIs since we cannot guarantee that messages received from third party packages will not have any unintended side-effects  on other Firebase products such as messaging delivery reporting and Analytics data.

To help us triage and locate genuine issues that need to be addressed we have created a specific issue template for iOS background messages. If you believe you still have an issue that needs to be addressed, please create a new issue following [this template](https://github.com/firebase/flutterfire/issues/new?assignees=&labels=Needs+Attention%2Ctype%3A+bug%2Cplugin%3A+messaging%2Cplatform%3A+ios&template=ios_background_message_handling.yml&title=%5Bfirebase_messaging%5D%3A+%3CYOUR+TITLE+HERE%3E).

I will be closing this issue in favor of raising a new issue with the new template above. This template will help you provide us with all the information we need to investigate a potential issue with background messaging on iOS.
